File: Lec8/Lec8.py
import queue
maxV = 100
INF = int(1e9)

class Node:
    def __init__(self, id, dist):
        self.w = dist
        self.v = id
    
    def __lt__(self, other):
        return self.w <= other.w
    
def dijkstra(s):
    pq = queue.PriorityQueue()
    pq.put(Node(s, 0))
    dist[s] = 0
    while not pq.empty():
        head = pq.get()
        if head.w != dist[head.v]:
            continue
        for adja in graph[head.v]:
            if head.w + adja.w < dist[adja.v]:
                dist[adja.v] = head.w + adja.w
                pq.put(Node(adja.v, dist[adja.v]))
                path[adja.v] = head.v

if __name__ == '__main__':    
    n = int(input())
    s, t = 0, 5
    graph = [[] for _ in range(n+3)]
    dist = [INF for _ in range(n+3)]
    path = [-1 for _ in range(n+3)]
    for i in range(n):
        ds = list(map(int, input().split()))
        for j in range(n):
            if ds[j]:
                graph[i].append(Node(j, ds[j]))
    dijkstra(s)
    ans = dist[t]
    print(ans)
    

File: Lec8/Lec8.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

#define maxV 100
const int INF = 1e9;

struct Tuple {
    int v;
    int w;

    Tuple(int vertex, int weight) : v(vertex), w(weight) {};
};

vector<vector<Tuple>> graph;
vector<int> dist(maxV, INF);
int path[maxV];

struct option {
    bool operator() (const Tuple& a, const Tuple& b) const {
        return a.w > b.w;
    }
};

void dijsktra(int s) {
    priority_queue<Tuple, vector<Tuple>, option> pq;
    pq.push(Tuple(s, 0));
    dist[s] = 0;
    while (!pq.empty()) {
        Tuple head = pq.top();
        pq.pop();
        if (dist[head.v] != head.w) continue;
        for (int i=0; i<graph[head.v].size(); i++) {
            Tuple adja = graph[head.v][i];
            if (head.w + adja.w < dist[adja.v]) {
                dist[adja.v] = head.w + adja.w;
                pq.push(Tuple(adja.v, dist[adja.v]));
                path[adja.v] = head.v;
            }
        }
    }   
}

int main() {
    int n, s, t;
    cin >> n;
    s = 0, t = 4;
    int d;
    for (int i=0; i<n; i++) {
        for (int j=0; j<n; j++) {
            cin >> d;
            if (d) graph[i].push_back(Tuple(j, d));
        }
    }
    dijsktra(s);
    int ans = dist[t];
    return 0;
}